### ISSUE TITLE
PRM evaluation stuffs

### ISSUE THREAD

**Issue by despzcm** [QwenLM/ProcessBench#3]
When you are evaluating PRM like Math-shepherd or skywork, i suppose the first thing to do is to evaluate scores of every steps. And how to set the threshold to determine the step correct or not according to the scores? 
Moreover, i find that you have only provided codes for evaluating critic models using majority voting generation. Could you provide codes to evaluate PRMs like Shepherd or Skywork?
Best regards

**Comment by chujiezheng** [comment]
Math-Shepherd and RLHFlow explicitly predict the correctness label for each step, so we directly use these predictions.

Skywork outputs a scalar score for each step. In the main experiments of the paper, we determine the threshold based on the F1 score on the GSM8K subset. In the Appendix, we also report the highest performance on each subset by respectively determining the optimal thresholds (*merely to see the upper bound performance* of Skywork PRMs).

I can upload the evaluation code for RLHFlow PRMs soon. You may either modify it to evaluate other PRMs, or await our release of code for Math-Shepherd and Skywork (but may be somewhat later).

**Comment by Gebro13** [comment]
Excellent, we understand you are picking the optimal thresholds due to the highest  F1-score in Skywork.
But, according to the Math-shepherd's releasing code, they actually also output a scalar score between 0 and 1 for each step by using softmax of + and - logits instead of directly predicting the correctness label. So would you consider it reasonable to pick the best threshold as is shown in Skywork?

**Comment by chujiezheng** [comment]
Personally I do not prefer it. Because:

1. We in principle should **NOT** tune the threshold on each subset, as this is actually tuning your hyperparameter on the test data. Making a precedent may stimulate future PRMs to overfit our benchmark.
2. I think a PRM (or any RM) should output the score in a stable range, to indicate the correctness of the step or something else meaningful. Outputting abnormal scores means the input becomes OOD to the RM, then the score would become less discriminative. For example, you can see that tuning the thresholds for the Skywork PRMs only bring marginal improvement, suggesting that they still fail to generalize and provide discriminative scores. Hence, I may expect that tuning the thresholds will not change the performance much.

For the above reasons, I think we should instead encourage future PRMs to be more robust, generalizable, and output scores that closely relate to the step correctness (so you can determine a threshold before any evaluation).

**Comment by Gebro13** [comment]
Thanks for your insightful response. That works a lot for me!

**Comment by despzcm** [comment]
Hi @chujiezheng 
Thank you so much for your valuable response! Your insights have been extremely helpful.

**Comment by chujiezheng** [comment]
@despzcm Please refer to this [code](https://github.com/QwenLM/ProcessBench/blob/main/code/run_eval_prm_rlhflow.py)
